src/util/mount.c

```diff
@@ -79,24 +82,50 @@ int check_mounted(char *mountpoint) {
         mountpoint[mountpoint_len-1] = '\0';
     }
 
-    real_mountpoint = realpath(mountpoint, NULL); // Flawfinder: ignore
-    if ( real_mountpoint == NULL ) {
-        // mountpoint doesn't exist
-        return(retval);
-    }
 
     singularity_message(DEBUG, "Iterating through /proc/mounts\n");
     while ( fgets(line, MAX_LINE_LEN, mounts) != NULL ) {
         (void) strtok(strdup(line), " ");
         char *mount = strtok(NULL, " ");
 
-        // Check to see if mountpoint is already mounted
-        if ( strcmp(joinpath(rootfs_dir, real_mountpoint), mount) == 0 ) {
-            singularity_message(DEBUG, "Mountpoint is already mounted: %s\n", mountpoint);
-            retval = 1;
-            break;
+        char *test_mountpoint = strdup(real_mountpoint);
+
+        while ( strcmp(test_mountpoint, "/") != 0 ) {
+            char *full_test_path = NULL;
+            char *tmp_test_path = joinpath(rootfs_dir, test_mountpoint);
+
+            if ( is_link(tmp_test_path) == 0 ) {
+                char *linktarget = realpath(tmp_test_path, NULL); // Flawfinder: ignore
+                if ( linktarget == NULL ) {
+                    singularity_message(ERROR, "Could not identify the source of contained link: %s\n", test_mountpoint);
+                    ABORT(255);
+                }
+                full_test_path = joinpath(rootfs_dir, linktarget);
+                singularity_message(DEBUG, "parent directory is a link, resolved: %s->%s\n", joinpath(rootfs_dir, test_mountpoint), full_test_path);
+                if ( strcmp(linktarget, "/") == 0 ) {
+                    retval = 1;
+                    free(test_mountpoint);
+                    goto DONE;
+                }
+                free(linktarget);
+            } else {
+                full_test_path = strdup(tmp_test_path);
+            }
+
+            // Check to see if mountpoint is already mounted
+            if ( strcmp(full_test_path, mount) == 0 ) {
+                singularity_message(DEBUG, "Mountpoint is already mounted: %s\n", test_mountpoint);
+                retval = 1;
+                free(test_mountpoint);
+                goto DONE;
+            }
+            test_mountpoint = dirname(test_mountpoint);
+            free(tmp_test_path);
+            free(full_test_path);
         }
 
+        free(test_mountpoint);
+
         // Check to see if path is in container root
         if ( strncmp(rootfs_dir, mount, strlength(rootfs_dir, 1024)) != 0 ) {
```

Comment:
I think this if and the following one don't do anything.  What's the point of doing "continue" at the end of a loop?